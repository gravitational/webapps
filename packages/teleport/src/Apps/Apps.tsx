/**
 * Copyright 2020-2022 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { Indicator, Box } from 'design';

import useTeleport from 'teleport/useTeleport';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import Empty, { EmptyStateInfo } from 'teleport/components/Empty';
import ErrorMessage from 'teleport/components/AgentErrorMessage';

import AgentButtonAdd from 'teleport/components/AgentButtonAdd';

import AppList from './AppList';
import useApps, { State } from './useApps';

export default function Container() {
  const ctx = useTeleport();
  const state = useApps(ctx);
  return <Apps {...state} />;
}

export function Apps(props: State) {
  const {
    clusterId,
    isLeafCluster,
    canCreate,
    attempt,
    results,
    fetchNext,
    fetchPrev,
    from,
    to,
    pageSize,
    params,
    setParams,
    startKeys,
    setSort,
    pathname,
    replaceHistory,
    fetchStatus,
    isSearchEmpty,
    onLabelClick,
  } = props;

  const hasNoApps = results.apps.length === 0 && isSearchEmpty;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center" justifyContent="space-between">
        <FeatureHeaderTitle>Applications</FeatureHeaderTitle>
        {attempt.status === 'success' && !hasNoApps && (
          <AgentButtonAdd
            agent="application"
            beginsWithVowel={true}
            isLeafCluster={isLeafCluster}
            canCreate={canCreate}
          />
        )}
      </FeatureHeader>
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'failed' && (
        <ErrorMessage message={attempt.statusText} />
      )}
      {attempt.status !== 'processing' && !hasNoApps && (
        <AppList
          apps={results.apps}
          fetchNext={fetchNext}
          fetchPrev={fetchPrev}
          fetchStatus={fetchStatus}
          from={from}
          to={to}
          totalCount={results.totalCount}
          pageSize={pageSize}
          params={params}
          setParams={setParams}
          startKeys={startKeys}
          setSort={setSort}
          pathname={pathname}
          replaceHistory={replaceHistory}
          onLabelClick={onLabelClick}
          paginationUnsupported={results.paginationUnsupported}
        />
      )}
      {attempt.status === 'success' && hasNoApps && (
        <Empty
          clusterId={clusterId}
          canCreate={canCreate && !isLeafCluster}
          emptyStateInfo={emptyStateInfo}
        />
      )}
    </FeatureBox>
  );
}

const emptyStateInfo: EmptyStateInfo = {
  title: 'Add your first application to Teleport',
  byline:
    'Teleport Application Access provides secure access to internal applications.',
  docsURL: 'https://goteleport.com/docs/application-access/getting-started/',
  resourceType: 'application',
  readOnly: {
    title: 'No Applications Found',
    resource: 'applications',
  },
};
