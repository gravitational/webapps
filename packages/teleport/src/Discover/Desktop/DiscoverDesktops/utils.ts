import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
