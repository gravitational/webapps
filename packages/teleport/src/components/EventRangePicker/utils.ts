import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
