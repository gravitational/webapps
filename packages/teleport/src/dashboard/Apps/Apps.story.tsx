/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import DefaultApps from './Apps';
import TeleportContextProvider from 'teleport/teleportContextProvider';
import TeleportContext from 'teleport/teleportContext';
import makeAcl from 'teleport/services/user/makeAcl';

export default {
  title: 'TeleportDashboard/Apps',
};

export const Loaded = () => {
  const ctx = new TeleportContext();
  const acl = makeAcl(sample.acl);

  ctx.storeUser.setState({ acl });
  ctx.appService.fetchApps = () => Promise.resolve(sample.apps);
  return render(ctx, <DefaultApps />);
};

export const Empty = () => {
  const ctx = new TeleportContext();
  const acl = makeAcl(sample.acl);

  ctx.storeUser.setState({ acl });
  ctx.appService.fetchApps = () => Promise.resolve([]);
  return render(ctx, <DefaultApps />);
};

export const Processing = () => {
  const ctx = new TeleportContext();
  const acl = makeAcl(sample.acl);

  ctx.storeUser.setState({ acl });
  ctx.appService.fetchApps = () => new Promise(() => null);
  return render(ctx, <DefaultApps />);
};

export const Failed = () => {
  const ctx = new TeleportContext();
  const acl = makeAcl(sample.acl);

  ctx.storeUser.setState({ acl });
  ctx.appService.fetchApps = () =>
    Promise.reject(new Error('some error message'));
  return render(ctx, <DefaultApps />);
};

function render(ctx: TeleportContext, children: JSX.Element) {
  return (
    <TeleportContextProvider value={ctx}>{children}</TeleportContextProvider>
  );
}

const sample = {
  acl: {
    apps: {
      list: true,
      create: true,
      remove: true,
      edit: true,
      read: true,
    },
  },
  apps: [
    {
      id: '1',
      name: 'jenkins',
      clusterId: 'root-cluster-name',
      labels: [],
      publicAddr: 'https://jekins.example.com/',
      internalAddr: 'localhost:8080',
      hostname: 'jenkin-hostname',
    },
    {
      id: '2',
      name: 'github',
      clusterId: 'leaf-cluster-name',
      labels: [
        { name: 'auth', value: 'dev' },
        { name: 'infrastructure', value: 'deployment' },
      ],
      publicAddr: 'https://github.example.com/',
      internalAddr: 'localhost:8081',
      hostname: 'github-hostname',
    },
    {
      id: '3',
      name: 'slack',
      clusterId: 'leaf2-cluster-longNameFormat-reallyLong',
      labels: [
        { name: 'auth', value: 'general' },
        { name: 'tools', value: 'comms' },
        { name: 'version', value: 'latest' },
        { name: 'teams', value: 'all' },
        { name: 'type', value: 'internal' },
      ],
      publicAddr: 'https://slack.example.com/',
      internalAddr: 'localhost:8080',
      hostname: 'slack-hostname',
    },
  ],
};
