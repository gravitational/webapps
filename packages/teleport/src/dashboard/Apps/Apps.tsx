/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Danger } from 'design/Alert';
import {
  Indicator,
  Text,
  Box,
  Flex,
  ButtonPrimary,
  ButtonSecondary,
} from 'design';
import Card from 'design/Card';
import Image from 'design/Image';
import AppListCards from './AppListCards';
import useApps from './useApps';

const aapPng = require('./application.png');

export default function Container() {
  const state = useApps();
  return <Apps {...state} />;
}

/**
 * Apps is a view component that based on permissions:
 * - Displays list of running apps
 * - Allows user to view details of an app
 */
export function Apps({ attempt, apps }: Props) {
  const isEmpty = attempt.status === 'success' && apps.length === 0;
  const hasApps = attempt.status === 'success' && !isEmpty;

  function onClickDoc() {
    window.open('https://gravitational.com/teleport/docs/', '_blank');
  }

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle>Applications</FeatureHeaderTitle>
        {hasApps && (
          <ButtonSecondary onClick={onClickDoc} ml="auto" width="240px">
            View Documentation
          </ButtonSecondary>
        )}
      </FeatureHeader>
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'failed' && <Danger>{attempt.statusText} </Danger>}
      {isEmpty && <Empty onClick={onClickDoc} />}
      {hasApps && <AppListCards apps={apps} />}
    </FeatureBox>
  );
}

const Empty = ({ onClick }) => {
  return (
    <Card maxWidth="700px" my={4} mx="auto" p={5} as={Flex} alignItems="center">
      <Box width={4 / 10}>
        <Image src={aapPng.default} width={'100%'} />
      </Box>
      <Box width={6 / 10} ml={5}>
        <Box mb={6}>
          <Text typography="h6" mb={3} caps>
            Secure Your First Application
          </Text>
          <Text typography="subtitle1" mb={3}>
            Teleport Application Access provides secure access to internal
            applications without the need for a VPN, but with the full control
            and audibility of Teleport.
          </Text>
        </Box>
        <ButtonPrimary onClick={onClick} width="240px">
          View Quickstart Guide
        </ButtonPrimary>
      </Box>
    </Card>
  );
};

type Props = ReturnType<typeof useApps>;
