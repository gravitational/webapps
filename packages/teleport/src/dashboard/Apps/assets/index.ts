const emptyPng = require('./empty.png');
export { emptyPng };
