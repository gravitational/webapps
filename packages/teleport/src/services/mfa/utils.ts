/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
