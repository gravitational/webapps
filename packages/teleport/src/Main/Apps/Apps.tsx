/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import { Danger } from 'design/Alert';
import {
  Indicator,
  Text,
  Box,
  Flex,
  ButtonPrimary,
  ButtonSecondary,
} from 'design';
import Card from 'design/Card';
import Image from 'design/Image';
import AppList from './AppList';
import useApps from './useApps';
import { emptyPng } from './assets';

const docUrl = 'https://gravitational.com/teleport/docs/';

export default function Container() {
  const state = useApps();
  return <Apps {...state} />;
}

export function Apps({ attempt, apps }: Props) {
  const isEmpty = attempt.status === 'success' && apps.length === 0;
  const hasApps = attempt.status === 'success' && !isEmpty;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle>Applications</FeatureHeaderTitle>
        {hasApps && (
          <ButtonSecondary
            as="a"
            href={docUrl}
            target="_blank"
            ml="auto"
            width="240px"
          >
            View Documentation
          </ButtonSecondary>
        )}
      </FeatureHeader>
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'failed' && <Danger>{attempt.statusText} </Danger>}
      {isEmpty && <Empty />}
      {hasApps && <AppList apps={apps} />}
    </FeatureBox>
  );
}

const Empty = () => {
  return (
    <Card maxWidth="700px" my={4} mx="auto" p={5} as={Flex} alignItems="center">
      <Box width={4 / 10}>
        <Image src={emptyPng} width={'100%'} />
      </Box>
      <Box width={6 / 10} ml={5}>
        <Box mb={6}>
          <Text typography="h6" mb={3} caps>
            Secure Your First Application
          </Text>
          <Text typography="subtitle1" mb={3}>
            Teleport Application Access provides secure access to internal
            applications without the need for a VPN, but with the full control
            and audibility of Teleport.
          </Text>
        </Box>
        <ButtonPrimary as="a" href={docUrl} target="_blank" width="240px">
          View Quickstart Guide
        </ButtonPrimary>
      </Box>
    </Card>
  );
};

type Props = ReturnType<typeof useApps>;
