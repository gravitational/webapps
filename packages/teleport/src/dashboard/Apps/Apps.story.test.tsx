import React from 'react';
import { Loaded, Viewing, Empty } from './Apps.story';
import { render } from 'design/utils/testing';

test('success state', async () => {
  const { container, findAllByText } = render(<Loaded />);
  const numCards = await findAllByText(/launch/i);
  expect(container).toMatchSnapshot();
  expect(numCards).toHaveLength(3);
});

test('empty state', async () => {
  const { container, findByText } = render(<Empty />);
  await findByText(/add your first application/i);
  expect(container).toMatchSnapshot();
});

test('viewing state', async () => {
  const { findByTestId } = render(<Viewing />);
  const container = await findByTestId('dialogbox');
  expect(container).toMatchSnapshot();
});
