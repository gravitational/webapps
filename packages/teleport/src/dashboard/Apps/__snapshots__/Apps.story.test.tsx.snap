// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty state 1`] = `
.c5 {
  box-sizing: border-box;
  width: 40%;
}

.c7 {
  box-sizing: border-box;
  width: 60%;
}

.c8 {
  box-sizing: border-box;
  margin-bottom: 40px;
  padding-right: 24px;
}

.c11 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  width: 240px;
}

.c11:active {
  opacity: 0.56;
}

.c11:hover,
.c11:focus {
  background: #651FFF;
}

.c11:active {
  background: #354AA4;
}

.c11:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c4 {
  box-sizing: border-box;
  max-width: 700px;
  margin-left: auto;
  margin-right: auto;
  margin-top: 24px;
  margin-bottom: 24px;
  padding-top: 24px;
  padding-bottom: 24px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c6 {
  display: block;
  outline: none;
  width: 100%;
}

.c9 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  font-size: 14px;
  line-height: 24px;
  margin: 0px;
  margin-bottom: 16px;
}

.c10 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 14px;
  line-height: 24px;
  margin: 0px;
  margin-bottom: 16px;
}

.c3 {
  box-sizing: border-box;
  max-width: 700px;
  margin-left: auto;
  margin-right: auto;
  margin-top: 24px;
  margin-bottom: 24px;
  padding-top: 24px;
  padding-bottom: 24px;
  background-color: #222C59;
  display: flex;
  align-items: center;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Apps
      </div>
    </div>
    <div
      class="c3 sc-AxirZ c4"
    >
      <div
        class="c5"
        width="0.4"
      >
        <img
          class="c6"
          width="100%"
        />
      </div>
      <div
        class="c7"
        width="0.6"
      >
        <div
          class="c8"
        >
          <div
            class="c9"
          >
            Add Your First Application
          </div>
          <div
            class="c10"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
          </div>
        </div>
        <button
          class="c11"
          kind="primary"
          width="240px"
        >
          Add Application
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`success state 1`] = `
.c9 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
  margin-top: auto;
  width: 100%;
}

.c9:active {
  opacity: 0.56;
}

.c9:hover,
.c9:focus {
  background: #651FFF;
}

.c9:active {
  background: #354AA4;
}

.c9:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c6 {
  display: inline-block;
  transition: color .3s;
  margin-top: 16px;
  margin-bottom: 16px;
  color: rgba(255,255,255,0.87);
  font-size: 48px;
}

.c7 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  font-size: 12px;
  line-height: 16px;
  text-transform: uppercase;
  margin: 0px;
  margin-bottom: 4px;
}

.c8 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
  color: rgba(255,255,255,0.87);
}

.c3 {
  box-sizing: border-box;
  display: flex;
  flex-wrap: wrap;
}

.c4 {
  box-sizing: border-box;
  margin-right: 32px;
  padding: 32px;
  height: 240px;
  width: 240px;
  background-color: #222C59;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
}

.c5 {
  box-sizing: border-box;
  margin-bottom: 16px;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  flex-shrink: 0;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Apps
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
        height="240px"
        style="position: relative; box-shadow: 0 8px 32px rgba(0, 0, 0, 0.24);"
        width="240px"
      >
        <div
          class="c5"
        >
          <span
            class="c6 icon icon-cloud "
            color="text.primary"
            font-size="48px"
          />
          <div
            class="c7"
          >
            jenkins
          </div>
          <div
            class="c8"
            color="text.primary"
          >
            root-cluster-name
          </div>
        </div>
        <button
          class="c9"
          kind="primary"
        >
          Launch
        </button>
      </div>
      <div
        class="c4"
        height="240px"
        style="position: relative; box-shadow: 0 8px 32px rgba(0, 0, 0, 0.24);"
        width="240px"
      >
        <div
          class="c5"
        >
          <span
            class="c6 icon icon-cloud "
            color="text.primary"
            font-size="48px"
          />
          <div
            class="c7"
          >
            github
          </div>
          <div
            class="c8"
            color="text.primary"
          >
            leaf-cluster-name
          </div>
        </div>
        <button
          class="c9"
          kind="primary"
        >
          Launch
        </button>
      </div>
      <div
        class="c4"
        height="240px"
        style="position: relative; box-shadow: 0 8px 32px rgba(0, 0, 0, 0.24);"
        width="240px"
      >
        <div
          class="c5"
        >
          <span
            class="c6 icon icon-cloud "
            color="text.primary"
            font-size="48px"
          />
          <div
            class="c7"
          >
            slack
          </div>
          <div
            class="c8"
            color="text.primary"
          >
            leaf2-cluster-longNameFormat-reallyLong
          </div>
        </div>
        <button
          class="c9"
          kind="primary"
        >
          Launch
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`viewing state 1`] = `
.c7 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c8 {
  box-sizing: border-box;
}

.c11 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  min-height: 32px;
  font-size: 12px;
  padding: 0px 24px;
}

.c11:active {
  opacity: 0.56;
}

.c11:hover,
.c11:focus {
  background: #2C3A73;
}

.c11:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c9 {
  box-sizing: border-box;
  border-radius: 100px;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  min-height: 16px;
  line-height: 1.4;
  font-size: 10px;
  font-weight: 500;
  padding: 0 8px;
  background-color: #111B48;
  color: rgba(255,255,255,0.87);
  margin-bottom: 4px;
  margin-right: 4px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  text-transform: uppercase;
  margin: 0px;
  color: rgba(255,255,255,0.87);
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 600;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
}

.c6 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 16px;
  display: flex;
}

.c0 {
  padding: 32px;
  padding-top: 24px;
  background: #1C254D;
  color: rgba(255,255,255,0.87);
  border-radius: 8px;
  box-shadow: 0 8px 32px rgba(0,0,0,0.24);
  display: flex;
  flex-direction: column;
  position: relative;
  overflow-y: auto;
  max-height: calc(100% - 96px);
  max-width: 500px;
  width: 100%;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 16px;
  min-height: 32px;
  display: flex;
  align-items: center;
}

.c3 {
  box-sizing: border-box;
  margin-bottom: 32px;
  flex: 1;
  display: flex;
  flex-direction: column;
}

.c10 {
  box-sizing: border-box;
}

<div
  class="c0"
  data-testid="dialogbox"
>
  <div
    class="c1"
  >
    <div
      class="c2"
      color="text.primary"
    >
      App Details
    </div>
  </div>
  <div
    class="c3"
  >
    <div
      class="c4"
    >
      <div
        class="c5"
        style="width: 126px;"
      >
        Application
        :
      </div>
      <div
        class="c6"
      >
        slack
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        style="width: 126px;"
      >
        Host name
        :
      </div>
      <div
        class="c6"
      >
        slack-hostname
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        style="width: 126px;"
      >
        Cluster Name
        :
      </div>
      <div
        class="c6"
      >
        leaf2-cluster-longNameFormat-reallyLong
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        style="width: 126px;"
      >
        Internal Address
        :
      </div>
      <div
        class="c6"
      >
        localhost:8080
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
        style="width: 126px;"
      >
        Public Address
        :
      </div>
      <div
        class="c6"
      >
        https://slack.example.com/
      </div>
    </div>
    <div
      class="c7"
    >
      <div
        class="c5"
      >
        Labels:
      </div>
      <div
        class="c8"
      >
        <div
          class="c9"
          kind="secondary"
        >
          auth: general
        </div>
        <div
          class="c9"
          kind="secondary"
        >
          tools: comms
        </div>
        <div
          class="c9"
          kind="secondary"
        >
          version: latest
        </div>
        <div
          class="c9"
          kind="secondary"
        >
          teams: all
        </div>
        <div
          class="c9"
          kind="secondary"
        >
          type: internal
        </div>
      </div>
    </div>
  </div>
  <div
    class="c10"
  >
    <button
      class="c11"
      kind="secondary"
    >
      Close
    </button>
  </div>
</div>
`;
